### PR TITLE
Detect cycles at config time and set PreserveReferences accordingly

### PR DESCRIPTION
**Pull request by lbargaoanu**
This affects the complex types benchmark. I guess I should modify it to prevent that. It might break some people, but I think we should remove Configuration.GetMappers().

### CODE REVIEW

**Comment by jbogard** [comment]
Ready to review?

**Comment by lbargaoanu** [comment]
Yes.

File: src/AutoMapper/Execution/TypeMapPlanBuilder.cs [hunk 1]
```diff
@@ -64,6 +66,64 @@ public LambdaExpression CreateMapperLambda()
...
+            if(propertyMap.SourceType == null)
+            {
+                return null;
+            }
+            var types = new TypePair(propertyMap.SourceType, propertyMap.DestinationPropertyType);
+            var typeMap = _configurationProvider.ResolveTypeMap(types);
+            if(typeMap == null)
+            {
+                var mapper = _configurationProvider.FindMapper(types) as IObjectMapperInfo;
+                if(mapper != null)
```
Comment (jbogard) [inline src/AutoMapper/Execution/TypeMapPlanBuilder.cs]: FYI you can do this as `if (_configurationProvider.FindMapper(types) is IObjectMapperInfo mapper)`. See https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#is-expression

**Comment by jbogard** [comment]
Do we know what GH issues this would close? I suspect a few.

File: src/AutoMapper/Execution/TypeMapPlanBuilder.cs (hunk 1 above)
Comment (lbargaoanu) [inline src/AutoMapper/Execution/TypeMapPlanBuilder.cs]: Yes, it was smth more compact in an earlier version, but for debugging, it was easier to write it more explicit.

**Comment by lbargaoanu** [comment]
#2068, unless you want UseDestinationValue before somebody requests it. But for other issues, the answer was to set PreserveReferences.

**Comment by lbargaoanu** [comment]
I guess we cannot deprecate PreserveReferences, it might be required in some cases, but we should let people know somehow that it's not usually required.

**Pull request merged**